From 801079bbca1faf28d3ab499f6a4f2f6f042a1ea7 Mon Sep 17 00:00:00 2001
From: Andrey Grodzovsky <andrey.grodzovsky@amd.com>
Date: Thu, 10 Oct 2019 16:41:54 +0800
Subject: [PATCH 0334/1453] drm/amdgpu: add lock for i2c bus

I2C bus must be locked at all time we access the I2C bus.

Signed-off-by: Andrey Grodzovsky <andrey.grodzovsky@amd.com>
Reviewed-by: Rui Teng <rui.teng@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/smu_v11_0_i2c.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/gpu/drm/amd/amdgpu/smu_v11_0_i2c.c b/drivers/gpu/drm/amd/amdgpu/smu_v11_0_i2c.c
index c44723c267c9..a3ebbc877031 100644
--- a/drivers/gpu/drm/amd/amdgpu/smu_v11_0_i2c.c
+++ b/drivers/gpu/drm/amd/amdgpu/smu_v11_0_i2c.c
@@ -632,6 +632,10 @@ static int smu_v11_0_i2c_eeprom_i2c_xfer(struct i2c_adapter *i2c_adap,
 	int i, ret;
 	struct amdgpu_ras_eeprom_control *control = to_eeprom_control(i2c_adap);
 
+#if !defined(HAVE_I2C_LOCK_OPERATIONS_STRUCT)
+	lock_bus(i2c_adap, 0);
+#endif
+
 	if (!control->bus_locked) {
 		DRM_ERROR("I2C bus unlocked, stopping transaction!");
 		return -EIO;
@@ -656,6 +660,10 @@ static int smu_v11_0_i2c_eeprom_i2c_xfer(struct i2c_adapter *i2c_adap,
 	}
 
 	smu_v11_0_i2c_fini(i2c_adap);
+
+#if !defined(HAVE_I2C_LOCK_OPERATIONS_STRUCT)
+	unlock_bus(i2c_adap, 0);
+#endif
 	return num;
 }
 
-- 
2.17.1

